### ISSUE TITLE
Question on CachingUserRepository

### ISSUE THREAD

**Issue by ade90036** [spring-projects/spring-data-examples#389]
I'm looking to reproduce the example provided in this repo, which is the spring data repository and spring caching via cache abstraction configured by annotations..  the example i'm referring to is the following: "CachingUserRepository.java".

I have been doing a lot of reading to understand what is going on and to be able to model it for my own needs.

Code in question is reproduced here for easy reference:

```
public interface CachingUserRepository extends CrudRepository<User, Long> {

	@Override
	@CacheEvict("byUsername")
	<S extends User> S save(S entity);

	@Cacheable("byUsername")
	User findByUsername(String username);
}
```

The question i have is regarding the "CacheEvict", where whenever a key is not specified the SimpleKeyGenerator is used to generate a key which is then used to evict the cache. 

However i don't understand how the eviction could ever possibly work given the cache key generated from the SimpleKeyGenerator will differ for the two method / cases.

I would suspect that have the same behaviour the CacheEvic method should also contain the following:

```
	@Override
	@CacheEvict(value="byUsername", key="#p0.username")
	<S extends User> S save(S entity);
```

Am i missing something beefy that spring is doing underneath?
Could someone if domain knowledge share some light on this?

Regards

ade

**Comment by mp911de** [comment]
Good catch, you're right, `@CacheEvict("byUsername")` uses `User` as key whereas `@Cacheable("byUsername")` on `findByUsername(…)` uses the user name property.

Do you want to submit a pull request to fix the issue or do you want us to fix it by changing the eviction annotation to `@CacheEvict(value="byUsername", key="#p0.username")`?

**Comment by ade90036** [comment]
Ok, i can manage a pull request. 
Do you have a document / policy for the pull request material?
ade

**Comment by ade90036** [comment]
hi... 

i have finally managed to get around to do this.
This is the pull request #412 

So i have made the code changes and tried to run the test.
However maven never picked up the test, this is because the test class is abstract.
Added a test for the eviction.

Please let me know if i screwed up the line spacing.. i have been carefull to not reformat anything. 

regards

ade

**Comment by mp911de** [comment]
Fixed with #412.
